adapter.go

```diff
@@ -319,3 +322,137 @@ func (a *Adapter) savePolicyLine(tx *ent.Tx, ptype string, rule []string) *ent.C
 
 	return line
 }
+
+// UpdatePolicy updates a policy rule from storage.
+// This is part of the Auto-Save feature.
+func (a *Adapter) UpdatePolicy(sec string, ptype string, oldRule, newPolicy []string) error {
+	return a.WithTx(func(tx *ent.Tx) error {
+		rule := a.toInstance(ptype, oldRule)
+		line := tx.CasbinRule.Update().Where(
+			casbinrule.PtypeEQ(rule.Ptype),
+			casbinrule.V0EQ(rule.V0),
+			casbinrule.V1EQ(rule.V1),
+			casbinrule.V2EQ(rule.V2),
+			casbinrule.V3EQ(rule.V3),
+			casbinrule.V4EQ(rule.V4),
+			casbinrule.V5EQ(rule.V5),
+		)
+		rule = a.toInstance(ptype, newPolicy)
+		line.SetV0(rule.V0)
+		line.SetV1(rule.V1)
+		line.SetV2(rule.V2)
+		line.SetV3(rule.V3)
+		line.SetV4(rule.V4)
+		line.SetV5(rule.V5)
+		_, err := line.Save(a.ctx)
+		return err
+	})
+}
+
+// UpdatePolicies updates some policy rules to storage, like db, redis.
+func (a *Adapter) UpdatePolicies(sec string, ptype string, oldRules, newRules [][]string) error {
+	return a.WithTx(func(tx *ent.Tx) error {
+		for _, policy := range oldRules {
+			rule := a.toInstance(ptype, policy)
+			if _, err := tx.CasbinRule.Delete().Where(
+				casbinrule.PtypeEQ(rule.Ptype),
+				casbinrule.V0EQ(rule.V0),
+				casbinrule.V1EQ(rule.V1),
+				casbinrule.V2EQ(rule.V2),
+				casbinrule.V3EQ(rule.V3),
+				casbinrule.V4EQ(rule.V4),
+				casbinrule.V5EQ(rule.V5),
+			).Exec(a.ctx); err != nil {
+				return err
+			}
+		}
+		lines := make([]*ent.CasbinRuleCreate, 0)
+		for _, policy := range newRules {
+			lines = append(lines, a.savePolicyLine(tx, ptype, policy))
+		}
+		if _, err := tx.CasbinRule.CreateBulk(lines...).Save(a.ctx); err != nil {
+			return err
+		}
+		return nil
+	})
+}
+
+// UpdateFilteredPolicies deletes old rules and adds new rules.
+func (a *Adapter) UpdateFilteredPolicies(sec string, ptype string, newPolicies [][]string, fieldIndex int, fieldValues ...string) ([][]string, error) {
+	oldPolicies := make([][]string, 0)
+	err := a.WithTx(func(tx *ent.Tx) error {
+		line := tx.CasbinRule.Query()
+		if fieldIndex <= 0 && 0 < fieldIndex+len(fieldValues) {
+			line = line.Where(casbinrule.V0EQ(fieldValues[0-fieldIndex]))
+		}
+		if fieldIndex <= 1 && 1 < fieldIndex+len(fieldValues) {
+			line = line.Where(casbinrule.V1EQ(fieldValues[1-fieldIndex]))
+		}
+		if fieldIndex <= 2 && 2 < fieldIndex+len(fieldValues) {
+			line = line.Where(casbinrule.V2EQ(fieldValues[2-fieldIndex]))
+		}
+		if fieldIndex <= 3 && 3 < fieldIndex+len(fieldValues) {
+			line = line.Where(casbinrule.V3EQ(fieldValues[3-fieldIndex]))
+		}
+		if fieldIndex <= 4 && 4 < fieldIndex+len(fieldValues) {
+			line = line.Where(casbinrule.V4EQ(fieldValues[4-fieldIndex]))
+		}
+		if fieldIndex <= 5 && 5 < fieldIndex+len(fieldValues) {
+			line = line.Where(casbinrule.V5EQ(fieldValues[5-fieldIndex]))
+		}
+		rules, err := line.All(a.ctx)
+		if err != nil {
+			return err
+		}
+		for _, rule := range rules {
+			if _, err := tx.CasbinRule.Delete().Where(
+				casbinrule.IDEQ(rule.ID),
```

Comment:
Do you think it's a better idea to check all fields of `rule`, I think it will prevent accidentally delete data updated by another session or transaction, also it will keep data consistency between `queryed` data and `returned` deleted data. (just checked the ent's `DeleteOne`'s behavior is same as delete by id.

Reply:
I didn't get it, but I guess the unique key would prevent it https://github.com/casbin/ent-adapter/blob/34b16dfbe080ff8d6e5a4c954675fd4dfc38c09e/ent/schema/casbinrule.go#L34

Reply:
I think the `unique` index only guarantee no data duplication. But you cant guarantee what happened between `query` and `delete`.

Reply:
@closetool

Reply:
@washanhanzi There's transaction in every method, it keeps atomic

Reply:
hmm, I don't think the default isolation level in mysql and postgresql will guarantee the atomic operation. Especially `ent` now didn't support `select for update`. I think the rows after the first `query` will not lock.

Reply:
Though I have no knowledge about how the adapter work with the enforcer.  I don't know is there any problems.

Reply:
@washanhanzi I will first merge this. You can send new issue and PR for this if there's still problem.